keps/sig-apps/19-Graduate-CronJob-to-Stable/README.md

```diff
@@ -0,0 +1,503 @@
+- Start Stop Tests
+  - Schedule cronjobs and randomly stop the controller and start it.
+  - Schedule cronjobs and stop the controller and start it after the deadline.
+
+### Conformance Tests
+
+The conformance tests are a subset of e2e tests. We will select test scenarios that we believe are expected from all conforming clusters. Then modify the test case to use the `framework.ConformanceIt()` function rather than the `framework.It()` function.
+These e2e tests shall be included in conformance tests:
+
+- ConcurrencyPolicy
+  - should schedule multiple jobs concurrently when AllowConcurrent
+  - should not schedule new jobs when ForbidConcurrent
+  - should replace jobs when ReplaceConcurrent
+- Suspend
+  - should not schedule jobs when suspended
+- SuccessfulJobsHistoryLimit
+  - should delete successful finished jobs when above successfulJobsHistoryLimit
+- FailedJobsHistoryLimit
+  - should delete failed finished jobs when above failedJobsHistoryLimit
+- StartingDeadlineSeconds
+  - Should schedule a job soon when missed the current window if StartingDeadlineSeconds is long
+- Tests covering Bug fixes
+  - [issue/63371 - Should not start “missed” jobs from old cronjob after updating time](https://github.com/kubernetes/kubernetes/issues/63371)
+- Start Stop Tests
+  - Schedule cronjobs and randomly stop the controller and start it.
+  - Schedule cronjobs and stop the controller and start it after the deadline.
+
+### Unit Tests
+
+This is subject to the new re-architected controller implementation. Overall these scenarios would be tested.
+
+- [Run or Not](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/pkg/controller/cronjob/cronjob_controller_test.go#L167) tests the controller under different scenarios to check if the Job is created or not
+- [Validates Job cleanup](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/pkg/controller/cronjob/cronjob_controller_test.go#L371) path of the controller under different conditions.
+- [Validates Status](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/pkg/controller/cronjob/cronjob_controller_test.go#L593) of the CronJob after sync under different conditions.
+
+## Implementation plan
+
+#### Release X
+
+- Alpha: Feature flag for new controller is disabled by default
+- Dual controller. Both old and new implementation co-exist
+
+#### Release X+1
+
+- Beta: Feature flag for new controller is enabled by default. If the distribution chooses it can be disabled
+- Introduce CronJob in batch/v1
+
+#### Release X+2
+
+- GA: The feature flag is deprecated and the old controller code is cleaned up
+
+## Graduation Criteria
+
+- [ ] Implement shared informers to reduce pressure on API Server
+- [ ] Pass conformance tests
+- [ ] Update documents reflecting the changes
+- [ ] Pass CronJob e2e tests
+- [ ] Pass CronJob unit-tests
+- [ ] Pass scale tests
+
```

Comment:
You also need to copy and fill in "Produciton Readiness" section:
https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template#production-readiness-review-questionnaire